src/storage/src/storage_state.rs

```diff
@@ -4,12 +4,67 @@
 // included in the LICENSE file.
 
 //! Worker-local state for storage timely instances.
+//!
+//! One instance of a [`Worker`], along with its contained [`StorageState`], is
+//! part of an ensemble of storage workers that all run inside the same timely
+//! cluster. We call this worker a _storage worker_ to disambiguate it from
+//! other kinds of workers, potentially other components that might be sharing
+//! the same timely cluster.
+//!
+//! ## Controller and internal communication
+//!
+//! A worker receives _external_ [`StorageCommands`](StorageCommand) from the
+//! storage controller, via a channel. Storage workers also share an _internal_
+//! control/command fabric ([`internal_control`](crate::internal_control)).
+//! Internal commands go through a `Sequencer` dataflow that ensures that all
+//! workers receive all commands in the same consistent order.
+//!
+//! We need to make sure that commands that cause dataflows to be rendered are
+//! processed in the same consistent order across all workers because timely
+//! requires this. To achieve this, we make sure that only internal commands can
+//! cause dataflows to be rendered. External commands (from the controller)
+//! cause internal commands to be broadcast (by only one worker), to get
+//! dataflows rendered.
+//!
+//! The internal command fabric is also used to broadcast messages from a local
+//! operator/worker to all workers. For example, when we need to tear down and
+//! restart a dataflow on all workers when an error is encountered.
+//!
+//! ## Async Storage Worker
+//!
+//! The storage worker has a companion [`AsyncStorageWorker`] that must be used
+//! when running code that requires `async`. This is needed because a timely
+//! main loop cannot run `async` code.
+//!
+//! ## Example flow of commands for `CreateSources`
+//!
+//! With external commands, internal commands, and the async worker,
+//! understanding where and how commands from the controller are realized can
+//! get complicated. We will follow the complete flow for `CreateSources`, as an
+//! example:
+//!
+//! 1. Worker receives a [`StorageCommand::CreateSources`] command from the
+//!    controller.
+//! 2. This command is processed in [`StorageState::handle_storage_command`].
+//! 3. During processing of that command, we call
+//!    [`AsyncStorageWorker::calculate_resume_upper`], which causes a command to
+//!    be sent to the async worker.
+//! 4. We eventually get a response from the async worker:
+//!    [`AsyncStorageWorkerResponse::IngestDescriptionWithResumeUpper`].
```

Comment:
Is it possible that we will apply controller commands in a different order than the controller sends them, given that some may be bounced off the async storage worker and others applied immediately? (Is the controller ordering ever important?)

Reply:
Yes! See this comment where I try and explain it: https://github.com/MaterializeInc/materialize/pull/16958/files?diff=split&w=0#r1072847940